File: 0322_CoinChange.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        sort(coins.begin(), coins.end(), greater<int>());
        if (coins.size() == 0) return -1;
        else if(amount ==0) return 0;
        
        int currentAmount = 0;
        int coinCount = 0;
        int idx = 0;
        while (true){
            if (currentAmount + coins[idx] < amount) {
                if(coinChange(coins, amount-coins[idx]) != -1){
                    currentAmount += coins[idx];
                    coinCount++;
                } else {
                    idx++;
                    if (idx == coins.size()){
                        coinCount = -1;
                        break;
                    }
                }
            } else if (currentAmount + coins[idx] == amount) {
                coinCount++;
                currentAmount = amount;
                break;
            } else {
                idx++;
                if (idx == coins.size()){
                    coinCount = -1;
                    break;
                }
            }
        }
        return coinCount;
    }
};

File: 0322_CoinChange.py
class Solution:
    def __init__(self):
        self.dic = {}
    def coinChange(self, coins: List[int], amount: int) -> int:
        if amount == 0:
            return 0
        elif amount < 0:
            return -1
        if amount in self.dic:
            return self.dic[amount]
        cnt = amount+1
        
        for c in coins:
            if amount-c == 0:
                cnt = 1
                break
            num = self.coinChange(coins, amount-c)
            if num > 0:
                if num+1 < cnt:
                    cnt = num+1
        if cnt == amount+1:
            cnt = -1
        self.dic[amount] = cnt
        return cnt
